### DISCUSSION TITLE
Export of an updated "photo.jpg" fails and generates "photo (1).jpg". Will it recover the original name on the next run?

### DISCUSSION THREAD

**Discussion by oPromessa** [RhetTbull/osxphotos#568] [Q&A] [answered]
Again thanks for the excellent work! Happy New Year!

I'm using osxphotos on Monterey to export into Synology NAS (into Synology Photos).

## Bumped into a issue and would like to understand the algorithm:
1. `photo.jpg`was exported on a previous osxphotos run.
2. An export of an updated `photo.jpg` fails and generates `photo (1).jpg` file and deletes the previous one `photo.jpg`.
3. Didn't rerun the export, but (here is my question) would it _recover_ and re/copy the file `photo.jpg` next time around and delete the `photo (1).jpg` file?

### Detailed sequence of events...
- From a previous run a photo already existed on destination.
- I believe osxphotos detected the need to copy the file and update the metadata: `_79A9889.JPG`
- Copy failed cause I'm using some "old disks" on the NAS (don't ask!... testing DSM 7.0 and osxphotos... will move to new disks shortly ). Will also start to use the `retry = 2` option. (_see below log_).
- But in the end a new file with a suffix `(1)` was created either by the copy or by exiftool `_79A9889 (1).JPG`
- As I'm using the `update` and `cleanup` options osxphotos then deleted the file `_79A9889.JPG` from the export folder (_see below log_).
- Although on the control database it still refers to the original filename (_see below log_).

### osxphotos output:
```
(...)
Exporting _79A9889.JPG (_79A9889.JPG) as _79A9889.JPG
2022-01-02 01:22:19,212 - WARNING - export_db.py - 222 - disk I/O error
2022-01-02 01:22:19,760 - WARNING - export_db.py - 553 - disk I/O error
2022-01-02 01:22:19,760 - WARNING - export_db.py - 398 - disk I/O error
Writing metadata with exiftool for /Volumes/photo-1/Family.Photos/2021-04-08 Spring Photoshoot Originals/_79A9889 (1).JPG
2022-01-02 01:22:21,032 - WARNING - export_db.py - 415 - disk I/O error
2022-01-02 01:22:21,033 - WARNING - export_db.py - 319 - disk I/O error
Exported new file /Volumes/photo-1/Family.Photos/2021-04-08 Spring Photoshoot Originals/_79A9889 (1).JPG
(...)
Deleting /Volumes/photo-1/Family.Photos/2021-04-08 Spring Photoshoot Originals/_79A9889.JPG
```

### osxphotos CSV report file contents:
```
filename,exported,new,updated,skipped,exif_updated,touched,converted_to_jpeg,sidecar_xmp,sidecar_json,sidecar_exiftool,missing,error,exiftool_warning,exiftool_error,extended_attributes_written,extended_attributes_skipped,cleanup_deleted_file,cleanup_deleted_directory,exported_album
/Volumes/photo-1/Family.Photos/2021-04-08 Spring Photoshoot Originals/_79A9889 (1).JPG,1,1,0,0,1,0,0,0,0,0,0,,,,0,0,0,0,
```

### osxphotos database on exported folder contents:
```
sqlite> select * from files  where filepath like "%_79A9889%";
3088|2021-04-08 Spring Photoshoot Originals/_79A9889.JPG|2021-04-08 spring photoshoot originals/_79a9889.jpg|585241C5-F80E-41EC-848E-0AA5DBA130EE|32768|4784196|1617879020.0|32768|4784278|1617879020.0
```

**Comment by RhetTbull** [comment, marked as answer]
Happy New Year to you as well! What's the full command line for your export? 

As the disk error was in the export database code which osxphotos uses to match filename to UUID (unique ID used by Photos to identify photos), it may be that the database lookup failed so the export code thought the existing file was a different UUID and created the new name with `(1)` to avoid collision.

Yes, if you rerun the export with --update, the original filename will be exported and --cleanup will delete the file with `(1)`

**Reply by RhetTbull** [reply]
I checked the code causing the disk I/O errors and confirmed what I wrote above.  In this case osxphotos functioned as designed.  

1. The export code saw a file with the target name already existed and checked the export database.
2. The target database read failed and in this case, the code traps the error and returns no results which is interpreted to mean the target file is not associated with the photo being exported and thus the new name with `(1)` appended is used. (For the code to reach this point, the file signatures must also have been different between source and target--this could be because the source was edited or you were using `--touch-file` or `--exiftool`)
3. The code attempted to write the new signature data back to the export database but this also failed with a disk I/O error.

If you re-run the code with `--cleanup` and `--update`, this file should get remedied back to the original, assuming there are no more disk I/O errors with the database.

If your target disk is flaky, `--retry` is probably a good idea but in this particular case, it wouldn't have solved the issue as `--retry` currently only retries the image export, not the database read/writes.  I'll open a new issue to add retry to database read/writes.

**Reply by oPromessa** [reply]
Super. I've just replaced the old disk with bad sectors and the RAID 5 is rebuilding. From Mac to NAS is a wired ethernet so there should be no more I/O Errors but adding the `--retry` to the db component would also be good.

Thanks again.

I'm doing copies step by step per year... but In a few days I'll do the huge copy of 80K pics from Photos into the NAS. 

I'm looking around at the code and at the features requests to see if I can help... will shout out if I find any part where I can be of help.

Thanks again!

**Reply by RhetTbull** [reply]
> I'm looking around at the code and at the features requests to see if I can help... will shout out if I find any part where I can be of help.

Always happy for help and PRs. I'm in the middle of a big refactor of the export code (which is a complex mess at the moment) so I'd stay away from there until I get the changes pushed. 

By the way, the [wiki](https://github.com/RhetTbull/osxphotos/wiki) might be helpful if you're looking to dive into the code or further explore the Photos database.

**Reply by oPromessa** [reply]
Further to this. Possibly not related with the tool but in case others have this issue. 


Now with a new disk on synology I expected not to have I/O errors. Ran `osxphotos`
- What happened was that synology immediately started converting the media files as they were copied which I believed overloaded the ds918+ and caused I/O errors. _it's an assumption only_
- Rescheduled the media files conversion on synology not to overlap with `osxphotos`copy operation and ran again. 
- Still I get and I/O error and a `database disk image is malformed` error. 
- But this time around I don't have any errors reported on the synology. 

My net from the Mac to synology does hop via a data switch (100Mbps), the main router (1Gbps) and then down another data switch (1Gbps) to reach the synology. So I'm puzzled on the reason for the error. 

**question: can I enable more debug to catch more details on the I/O error?**

Other than that I can try a direct cable just to troubleshoot. 

```
Exporting IMG_20210718_142743.JPG (IMG_20210718_142743.JPG) as IMG_20210718_142743.JPG
Writing metadata with exiftool for /Volumes/photo-1/Family.Photos/2021-07-12 Trip to Madrid/IMG_20210718_142743.JPG
Exported new file /Volumes/photo-1/Family.Photos/2021-07-12 Trip to Madrid/IMG_20210718_142743.JPG
Exporting IMG_20210629_185343.JPG (IMG_20210629_185343.JPG) as IMG_20210629_185343.JPG
2022-01-03 13:45:03,322 - WARNING - export_db.py - 553 - disk I/O error
2022-01-03 13:45:03,322 - WARNING - export_db.py - 398 - database disk image is malformed

```

**Reply by oPromessa** [reply]
Maybe if the .osxphotos database file would be a local file on the Mac - thus avoiding the errors - and could then be moved to the expected directory at the end.

**Reply by RhetTbull** [reply]
> question: can I enable more debug to catch more details on the I/O error?

There is a debug mode (--debug) but it mostly just dumps out data that is parsed out of the Photos database. I don't think the export DB code (export_db.py) has any debug enabled.  The errors being posted are the exceptions being caught be the sqlite driver so I'm not sure how much more info you can get out of these.

You could clone the repo and add some instrumentation or breakpoints to export_db.py to try to get some additional debug info.

I've not seen this issue before, even with flaky disks.  When you changed the disk were you updating an existing export or running a new export?  If updating, it's possible the database got corrupted while it was on the bad disk.  You could try deleting (or moving) the database and re-running with --update.  If you didn't use --exiftool or --touch-file the update process will rebuild the export database and should be able to do so without re-exporting most of the photos.  If you did use --exiftool or --touch-file, the signatures of the exported files won't match what's in Photos and they'll be re-exported.

Another way to test whether it's a corrupt database might be to try the same export but with --dry-run.  This will cause osxphotos to load the export database into memory and all operations on the database will happen against an in-memory version.  If the on-disk version is corrupted I expect you'd get errors with the in-memory version too.

> Maybe if the .osxphotos database file would be a local file on the Mac - thus avoiding the errors - and could then be moved to the expected directory at the end.

You can specify a different name for the export database but not a different path as currently designed.  The current design uses relative paths to the export database to track photos so moving the database won't work.  It would take some re-work to make absolute paths work and this gets tricky for the end-user as the database is now potentially separate from the exported images.

**Reply by RhetTbull** [reply]
I double checked the export_db.py code to confirm this issue and unfortunately, path of export DB does need to be relative to the export directory:

https://github.com/RhetTbull/osxphotos/blob/147b30f97308db65868dc7a8d177d77ad0d0ad40/osxphotos/export_db.py#L208-L212

It would be possible to rework this but would need to think about how to do this in a backwards compatible way. Keep the paths relative allows easily moving the entire export tree and having update still work. I'm not sure it's worth the work of reworking this just for the sake of a flaky edge case.

**Reply by RhetTbull** [reply]
@oPromessa After looking at the code for the ExportDB I realized this wouldn't be too hard to implement so I went ahead and did it. I changed ExportDB() to require path to the export directory which it now uses to compute relative paths instead of its parent directory. You can now specify the export database to reside anywhere, including another volume.  osxphotos will print a warning though if export database is not in the same path as the export destination.  See version [0.44.4](https://github.com/RhetTbull/osxphotos/releases/tag/v0.44.4).

**Reply by oPromessa** [reply]
Tested it out. 
- with files on both locations. Got the warning. Working well.
- with only one file. Working well also.
